File: 0453_Minimum_Moves_to_Equal_Array_Elements/solution.py
'''
453. Minimum Moves to Equal Array Elements

Given a non-empty integer array of size n, find the minimum number of moves required to make all array
elements equal, where a move is incrementing n - 1 elements by 1.

Example:

Input:
[1,2,3]

Output:
3

Explanation:
Only three moves are needed (remember each move increments two elements):

[1,2,3]  =>  [2,3,3]  =>  [3,4,3]  =>  [4,4,4]

'''

# http://bookshadow.com/weblog/2016/11/06/leetcode-minimum-moves-to-equal-array-elements/
class Solution(object):
    def minMoves(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return sum(nums) - min(nums) * len(nums)

File: 0453_Minimum_Moves_to_Equal_Array_Elements/solution.java
package cn.huang.leetcode;

/*
453. Minimum Moves to Equal Array Elements

 */
public class LC_0453_MinimumMovestoEqualArrayElements {
    /*
    每次需要给除了数组最大值的所有数字加1

    全部n-1个值加1就是一个值减1
     */
    public int minMoves(int[] nums) {
        int sum = 0, len = nums.length;
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < len; i++) {
            sum += nums[i];
            if (nums[i] < min)
                min = nums[i];
        }
        return sum - len * min;

    }
}
